File: br/Lib/LretUtil.cs
//ルーレット処理ライブラリ
using System;
using System.Collections.Generic;
using System.Linq;
using Utils;
using XorShifts;

namespace LretUtils{
	//ルーレット設定用オブジェクト
	class PlayerItems{
		public int index=-1;
		//ルーレット対象リスト
		public string[] values;
		public string value{get{return values[index];}}
		//空白埋めリスト
		private string[] spaces;
		public string space{get{return spaces[index];}}

		//コンストラクタ
		public PlayerItems(string[] values){
			this.values=values;
			//空白埋めリストの生成
			int[] valuesLen=Array.ConvertAll(values,s=>Util.lenB(s));
			int valuesLenMax=valuesLen.Max();
			spaces=Array.ConvertAll(valuesLen,i=>Util.padStr("",i,valuesLenMax));
		}
	}

	//プレイヤークラス
	class Player{
		public string name;
		public Dictionary<string,PlayerItems> items;

		//ルーレット結果
		public string result{get{
			var res=name;
			var resSpace="";
			foreach(var (key,v) in items){
				if(0<=v.index){
					res+=$@"{resSpace}　{key}:{v.value}";
					resSpace=v.space;
				}
			}
			return res;
		}}

		//ルーレット結果(twitter)
		public string resultTwit{get{
			string res=name;
			foreach(var (key,v) in items){
				if(0<=v.index){
					res+=$@"　{key}:{v.value}";
				}
			}
			return res;
		}}

		//コンストラクタ
		public Player(string name,Dictionary<string,string[]> items){
			this.name=name;
			//与えられたリストをオブジェクト化
			this.items=items.ToDictionary(v=>v.Key,v=>new PlayerItems(v.Value));
		}
	}

	//ルーレット用関数
	static class Lret{
		//プレイヤー項目用ルーレット
		public static void item(
			List<Player> players, //プレイヤーリスト
			string[] lasts,       //プレイヤー毎の履歴リスト
			string[] itemList,    //ルーレット対象
			string key,           //ルーレット項目
			bool isOptionAll,     //項目の統一を行う?
			bool isOptionOnly,    //項目の重複をさせない?
			XorShift xs           //乱数ジェネレータ
		){
			//ルーレットが失敗した時もう一度
			for(;;){
			repeat:
				//履歴はルーレット対象は最低2つ以上存在する場合のみ参照する
				var isLastChk=2<=itemList.Length;

				//項目の統一させる場合、
				if(isOptionAll){
					int x=xs.randInt(0,itemList.Length-1);
					//全てのプレイヤーに等しく結果を代入
					//履歴は1人目のプレイヤーの物のみ参照する
					foreach(var (p,i) in players.Select((p,i)=>(p,i))){
						if(isLastChk && i==0 && lasts[0]==itemList[x]) goto repeat;
						p.items[key].index=x;
					}
					//履歴を更新
					for(int i=0;i<lasts.Length;i++) lasts[i]=players[0].items[key].value;
				}
				else{
					int[] xList;
					//項目の重複をさせない場合
					if(isOptionOnly && players.Count<itemList.Length){
						//重複しない乱数で結果を生成
						xList=Enumerable.Range(0,itemList.Length).ToArray();
						xList=xs.shuffle(xList);
					}
					//通常条件
					else{
						//通常の乱数
						xList=Enumerable.Range(0,players.Count)
							.Select(v=>xs.randInt(0,itemList.Length-1)).ToArray();
					}
					if(isLastChk){
						//履歴を参照して2連続で選択されるものがあればもう一回
						foreach(var (p,i) in players.Select((p,i)=>(p,i))){
							if(lasts[i]==itemList[xList[i]]) goto repeat;
						}
					}
					//ルーレット結果をプレイヤーに代入、履歴を更新
					foreach(var (p,i) in players.Select((p,i)=>(p,i))){
						p.items[key].index=xList[i];
						lasts[i]=p.items[key].value;
					}
				}
				return;
			}
		}

		//クエストルーレット
		public static string quest(List<string> questLasts,List<string> questList,XorShift xs){
			//履歴保存数
			int n;
			if(questLasts.Count==0 || questList.Count<=2)
				n=0;
			else if(questList.Count-1<=questLasts.Count)
				n=questList.Count-2;
			else if(questLasts.Count<10)
				n=questLasts.Count;
			else
				n=10;

			//履歴数調整
			questLasts.RemoveRange(0,questLasts.Count-n);

			//履歴に含まれるクエストは選択しない
			string quest;
			do{
				quest=questList[xs.randInt(0,questList.Count-1)];
			}while(2<=questList.Count && questLasts.Contains(quest));
			questLasts.Add(quest);

			return quest;
		}
	}
}

File: ve/src/lib/LretUtil.ts
//ルーレット処理ライブラリ
import Util from "@/lib/Util";
import {XorShift} from "@/lib/XorShift";
import Item from "@/lib/Item";

//ルーレット設定用オブジェクト
class PlayerItems{
	public index=-1;
	//ルーレット対象リスト
	public values:string[];
	public get value():string{return this.values[this.index];}
	//空白埋めリスト
	private spaces:string[];
	public get space():string{return this.spaces[this.index];}

	//コンストラクタ
	constructor(values:string[]){
		this.values=values;
		//空白埋めリストの生成
		let valuesLen=values.map(s=>Util.lenB(s));
		let valuesLenMax=Math.max(...valuesLen);
		this.spaces=valuesLen.map(i=>Util.padStr("",i,valuesLenMax));
	}
}

//プレイヤークラス
export class Player{
	public name:string;
	public items:{[s:string]:PlayerItems}={};

	//ルーレット結果
	public get result():string{
		var res=this.name;
		var resSpace="";

		for(let [key,v] of Object.entries(this.items)){
			if(0<=v.index){
				res+=`${resSpace}　${key}:${v.value}`;
				resSpace=v.space;
			}
		}
		return res;
	}

	//ルーレット結果(twitter)
	public get resultTwit():string{
		var res=this.name;

		for(let [key,v] of Object.entries(this.items)){
			if(0<=v.index){
				res+=`　${key}:${v.value}`;
			}
		}
		return res;
	}

	//コンストラクタ
	constructor(name:string,items:{[s:string]:string[]}){
		this.name=name;
		//与えられたリストをオブジェクト化
		Object.entries(items).forEach(([key,value])=>this.items[key]=new PlayerItems(value));
	}
}

//ルーレット用関数
export class Lret{
	//プレイヤー項目用ルーレット
	public static item(       //プレイヤーリスト
		players:Player[],     //プレイヤー毎の履歴リスト
		lasts:string[],       //ルーレット対象
		itemList:string[],    //ルーレット項目
		key:string,           //項目の統一を行う?
		isOptionAll:boolean,  //項目の重複をさせない?
		isOptionOnly:boolean, //乱数ジェネレータ
		xs:XorShift
	){
		//ルーレットが失敗した時もう一度
		repeat:
		for(;;){
			//履歴はルーレット対象は最低2つ以上存在する場合のみ参照する
			let isLastChk=2<=itemList.length;

			//項目の統一させる場合、
			if(isOptionAll){
				let x=xs.randInt(0,itemList.length-1);
				//全てのプレイヤーに等しく結果を代入
				//履歴は1人目のプレイヤーの物のみ参照する
				for(let [i,p] of players.entries()){
					if(isLastChk && i==0 && lasts[0]==itemList[x]) continue repeat;
					p.items[key].index=x;
				}
				//履歴を更新
				for(let i=0;i<lasts.length;i++) lasts[i]=players[0].items[key].value;
			}
			else{
				let xList:number[];
				//項目の重複をさせない場合
				if(isOptionOnly && players.length<itemList.length){
					//重複しない乱数で結果を生成
					xList=[...itemList.keys()];
					xList=xs.shuffle(xList);
				}
				//通常条件
				else{
					//通常の乱数
					xList=[...players.keys()]
						.map(v=>xs.randInt(0,itemList.length-1));
				}
				if(isLastChk){
					//履歴を参照して2連続で選択されるものがあればもう一回
					for(let [i,p] of players.entries()){
						if(lasts[i]==itemList[xList[i]]) continue repeat;
					}
				}
				//ルーレット結果をプレイヤーに代入、履歴を更新
				for(let [i,p] of players.entries()){
					p.items[key].index=xList[i];
					lasts[i]=p.items[key].value;
				}
			}
			return;
		}
	}

	//クエストルーレット
	public static quest(questLasts:string[],questList:string[],xs:XorShift):string{
		//履歴保存数
		var n:number;
		if(questLasts.length==0 || questList.length<=2)
			n=0;
		else if(questList.length-1<=questLasts.length)
			n=questList.length-2;
		else if(questLasts.length<10)
			n=questLasts.length;
		else
			n=10;

		//履歴数調整
		questLasts.splice(0,questLasts.length-n);

		//履歴に含まれるクエストは選択しない
		var quest:string;
		do{
			quest=questList[xs.randInt(0,questList.length-1)];
		}while(2<=questList.length && questLasts.includes(quest));
		questLasts.push(quest);

		return quest;
	}
}
